core/src/main/java/org/elasticsearch/search/DocValueFormat.java

```diff
@@ -124,14 +126,20 @@ public BytesRef parseBytesRef(String value) {
 
         public static final String NAME = "date_time";
 
-        final FormatDateTimeFormatter formatter;
+        final String format;
         final DateTimeZone timeZone;
         private final DateMathParser parser;
+        private final DateTimeFormatter formatter;
 
         public DateTime(FormatDateTimeFormatter formatter, DateTimeZone timeZone) {
-            this.formatter = Objects.requireNonNull(formatter);
+            this.format = Objects.requireNonNull(formatter.format());
             this.timeZone = Objects.requireNonNull(timeZone);
             this.parser = new DateMathParser(formatter);
+            if (formatter.format().startsWith("epoch_millis") || formatter.format().startsWith("epoch_second")) {
+                this.formatter = formatter.printer();
```

Comment:
I looked into this some more. It's problematic, the actual joda EpochPrinter prints whatever time instant (long) is given to it, the time zone adjustment happens in joda DateTimeFormatter.printTo(Appendable, long, Chronology), and if we use `withZone()` (which we need for other formatters to work correct) the time zone offset gets added. I don't see another way around this atm. but am open to suggestions where to dig further if you have ideas.